File: src/main/java/com/example/dpop/orchestrator/api/v1/OrchestratorResponse.java
package com.example.dpop.orchestrator.api.v1;

import com.fasterxml.jackson.annotation.JsonInclude;

import java.util.List;
import java.util.UUID;

@JsonInclude(JsonInclude.Include.NON_NULL)
public record OrchestratorResponse(
        UUID channelSessionId,
        ProcessState processState,
        AttemptState attemptState,
        NextRouting next,
        DemoHints _demo
) {
    public OrchestratorResponse(UUID channelSessionId, NextRouting next) {
        this(channelSessionId, null, null, next, null);
    }

    public OrchestratorResponse(UUID channelSessionId, ProcessState processState, NextRouting next) {
        this(channelSessionId, processState, null, next, null);
    }

    public OrchestratorResponse(UUID channelSessionId, ProcessState processState, AttemptState attemptState, NextRouting next) {
        this(channelSessionId, processState, attemptState, next, null);
    }

    /** Only present in demo/test mode — never included in production responses. */
    @JsonInclude(JsonInclude.Include.NON_NULL)
    public record DemoHints(
            String tan,
            String note
    ) {}

    @JsonInclude(JsonInclude.Include.NON_NULL)
    public record ProcessState(
            String purpose,
            String status,
            Long personId,
            Long accountId
    ) {}

    @JsonInclude(JsonInclude.Include.NON_NULL)
    public record AttemptState(
            UUID attemptId,
            String attemptType,
            String status,
            List<String> missingFields,
            Object result
    ) {}

    @JsonInclude(JsonInclude.Include.NON_NULL)
    public record NextRouting(
            String context,
            String step,
            List<String> methods,
            String enrollmentRef,
            Long accountId,
            Long personId
    ) {
        public NextRouting(String context, String step) {
            this(context, step, null, null, null, null);
        }

        public NextRouting(String context, String step, List<String> methods) {
            this(context, step, methods, null, null, null);
        }

        public NextRouting(String context, String step, List<String> methods, String enrollmentRef) {
            this(context, step, methods, enrollmentRef, null, null);
        }
    }
}

File: src/main/kotlin/com/example/dpop/orchestrator/api/v1/OrchestratorResponse.kt
package com.example.dpop.orchestrator.api.v1

import com.fasterxml.jackson.annotation.JsonInclude
import java.util.UUID

@JsonInclude(JsonInclude.Include.NON_NULL)
@JvmRecord
data class OrchestratorResponse(
    val channelSessionId: UUID?,
    val processState: ProcessState?,
    val attemptState: AttemptState?,
    val next: NextRouting?,
    val _demo: DemoHints?
) {
    constructor(channelSessionId: UUID?, next: NextRouting?) :
        this(channelSessionId, null, null, next, null)

    constructor(channelSessionId: UUID?, processState: ProcessState?, next: NextRouting?) :
        this(channelSessionId, processState, null, next, null)

    constructor(
        channelSessionId: UUID?,
        processState: ProcessState?,
        attemptState: AttemptState?,
        next: NextRouting?
    ) : this(channelSessionId, processState, attemptState, next, null)

    /** Only present in demo/test mode — never included in production responses. */
    @JsonInclude(JsonInclude.Include.NON_NULL)
    @JvmRecord
    data class DemoHints(
        val tan: String?,
        val note: String?
    )

    @JsonInclude(JsonInclude.Include.NON_NULL)
    @JvmRecord
    data class ProcessState(
        val purpose: String?,
        val status: String?,
        val personId: Long?,
        val accountId: Long?
    )

    @JsonInclude(JsonInclude.Include.NON_NULL)
    @JvmRecord
    data class AttemptState(
        val attemptId: UUID?,
        val attemptType: String?,
        val status: String?,
        val missingFields: List<String>?,
        val result: Any?
    )

    @JsonInclude(JsonInclude.Include.NON_NULL)
    @JvmRecord
    data class NextRouting(
        val context: String?,
        val step: String?,
        val methods: List<String>?,
        val enrollmentRef: String?,
        val accountId: Long?,
        val personId: Long?
    ) {
        constructor(context: String?, step: String?) :
            this(context, step, null, null, null, null)

        constructor(context: String?, step: String?, methods: List<String>?) :
            this(context, step, methods, null, null, null)

        constructor(context: String?, step: String?, methods: List<String>?, enrollmentRef: String?) :
            this(context, step, methods, enrollmentRef, null, null)
    }
}
